include/sacn/merge.h

```diff
@@ -0,0 +1,162 @@
+ *     source of the merged data value, or that no source currently owns this address.
+ *
+ * \TODO: Add sample usage
+ *
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*! Each universe has a handle associated with it.*/
+typedef int universe_handle_t;
+
+/*! The sources on a universe have a short id that is used in the owned values, rather than a UUID.*/
+typedef uint16_t source_id_t;
+
+/*! An invalid source id handle value. */
+#define DMX_MERGER_SOURCE_INVALID -1
+
+/*! A set of configuration information for a universe to be merged. */
+typedef struct DmxMergerUniverseConfig
+{
+  /*! The maximum number of sources this universe will listen to.  May be #SACN_RECEIVER_INFINITE_SOURCES.
+      This parameter is ignored when configured to use static memory -- #DMX_MERGER_MAX_SOURCES_PER_UNIVERSE is used
+      instead.*/
+  size_t source_count_max;
+
+  /*! Buffer of DMX_ADDRESS_COUNT levels that this library keeps up to date as it merges.
+      Memory is owned by the application.*/
+  uint8_t* slots;
+
+  /*! Buffer of DMX_ADDRESS_COUNT source IDs that indicate the current winner of the merge for
+      that slot, or DMX_MERGER_SOURCE_INVALID to indicate that no source is providing values for that slot.
+      Memory is owned by the application.*/
+  source_id_t* slot_owners;
+
+} DmxMergerUniverseConfig;
+
+/*!
+ * \brief An initializer for an DmxMergerUniverseConfig struct.
+ *
+ * Usage:
+ * \code
+ * // Create the struct
+ * DmxMergerUniverseConfig universe_config = DMX_MERGER_UNIVERSE_CONFIG_INIT;
+ * // Now fill in the members of the struct
+ * \endcode
+ *
+ */
+#define DMX_MERGER_UNIVERSE_CONFIG_INIT \
+  {                                     \
+    0, NULL, NULL                       \
+  }
+
+/*! The current input data for a single source of the merge.
+    This is exposed only for informational purposes, as the
+    application call a variant of *TODO* */
+typedef struct DmxMergerSource
+{
+  /*! The UUID (e.g. sACN CID) of the DMX source. */
+  EtcPalUuid cid;
+
+  /*! The DMX data values (0 - 255). */
+  uint8_t values[DMX_ADDRESS_COUNT];
+
+  /*! Some sources don't send all 512 values, so here's how much of values to use.*/
+  size_t valid_value_count;
+
+  /*! The sACN per-universe priority (0 - 255). */
+  uint8_t universe_priority;
+
+  /*! Whether or not the address_priority buffer is valid. */
+  bool address_priority_valid;
+
+  /*! The sACN per-address (startcode 0xdd) priority (1-255, 0 means not sourced).
+      If the source does not */
+  uint8_t address_priority[DMX_ADDRESS_COUNT];
+
+} DmxMergerSource;
+
+etcpal_error_t dmx_merger_init();
+void dmx_merger_deinit(void);
+
+etcpal_error_t dmx_merger_create_universe(const DmxMergerUniverseConfig* config, universe_handle_t* handle);
+etcpal_error_t dmx_merger_destroy_universe(universe_handle_t handle);
+
+etcpal_error_t dmx_merger_add_source(universe_handle_t universe, const EtcPalUuid* source_cid, source_id_t* source_id);
```

Comment:
So does this API wrap the sACN Receiver API, or is the user expected to add sources manually using this function when a new source is delivered via the universe_data callback?

Reply:
It does not wrap the Receiver API.  You're right that there's currently an impedance mismatch.  I am planning on changing the Receiver API to have a SourcesFound callback rather than relying on the UniverseData callback to cause source creation.  Also, I wanted users to be able to add sources for other inputs.

My goal is to have the outputs of the Receiver API be trivial to pipe to this API, so you're right that we'll want to keep these issues in mind.

Reply:
OK, I get it. In that case I think we should have another API, which links them together, for maximum ease of use.

Reply:
Basically I want the all-software sACN merge case (e.g. network fixtures) to be able to use a single API and not have to hook two together.

Reply:
I chose this method so that the application can choose what to do with the received sACN output: e.g. give it to the FPGA or give it to the merger.

The merger currently doesn't do anything asynchronous.  If the merger was able to register callbacks with the receiver library, would it then need an notification so the app knows the values have changed, or would the app free wheel over the buffer?

Reply:
I would imagine that being the job of the yet-unnamed third, encompassing API.

So for example:

```
  sACN Merge Receiver
    /             \
sACN Receiver  DMX Merger
```

DMX Merger would remain all-synchronous.
Apps that use hardware merging would use sACN receiver only.
Apps that merge sACN with other DMX sources would use sACN Receiver and DMX Merger, hooking them up manually.
Apps that only need software-merged sACN could use sACN Merge Receiver.

Reply:
I see your point, but I hesitate.  Would it be ok for us to see how these libraries go before specifying and reviewing requirements/api for an additional library?  If the API is nearly a pass-through, we could provide examples for the sACN receiver APIs that show the needed call to the merger.

Reply:
Oh, I wasn't thinking of an additional library, just another header/module in this library.

Reply:
Fair enough, I've noticed that I tend to conflate the two.  But it still would mean requirements & review.

Reply:
I guess I don't foresee it being an especially contentious review item? Where is your hesitation coming from?

Reply:
Two mental blocks come to mind, both of which may be completely invalid:
1) This feels like ACNConnect, and that was added to cover app-unfriendliness in the lower library.  I'm worried that we're adding unnecessary layers.
2) I also have a feeling that applications may want to use this data differently (e.g. a notification on every change vs. freewheeling on the buffers).